### DISCUSSION TITLE
Using Multiprocessing combined with Casadi

### DISCUSSION THREAD

**Discussion by gregoriomarchesini** [casadi/casadi#3597] [Q&A]
Hey Developers!

Thank you very much for the outstanding development of the tool and the support you provide. 
Recently I have been trying to use a multiprocessing toolbox with Casadi. I tried different settings but I am not able to get multiprocessing running with a simple Casadi optimization program. This is due to the `pickle` module mainly.

I give here a simple example that would be useful for me to run :

```
from casadi import *
from multiprocessing import Pool


    

class ParSolver:
    def __init__(self) -> None:
       
        self.x = MX.sym("x")
        self.solver = nlpsol("solver","ipopt",{"x":self.x,"f":sin(self.x)**2})
    
    def _solve_single(self,x0):
        res = self.solver(x0 = x0)
        return res["x"]
    
    def solve(self,samples:list[float]):
        
        pool = Pool(4)
        res = pool.starmap(self._solve_single,samples)
        
        return min(res)

pp = ParSolver()
pp.solve([1,2,3,4,5,4,5,6,6,7,8,9,9])
```

If I try to run it, I get the following  error :
```
Traceback (most recent call last):
  File "/home/gregorio/Desktop/advanced_dynamics_project/leafSTL_analysis/stupid copy.py", line 26, in <module>
    pp.solve([1,2,3,4,5,4,5,6,6,7,8,9,9])
  File "/home/gregorio/Desktop/advanced_dynamics_project/leafSTL_analysis/stupid copy.py", line 20, in solve
    res = pool.starmap(self._solve_single,samples)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gregorio/miniforge3/envs/hetero_dyn/lib/python3.12/multiprocessing/pool.py", line 375, in starmap
    return self._map_async(func, iterable, starmapstar, chunksize).get()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gregorio/miniforge3/envs/hetero_dyn/lib/python3.12/multiprocessing/pool.py", line 774, in get
    raise self._value
  File "/home/gregorio/miniforge3/envs/hetero_dyn/lib/python3.12/multiprocessing/pool.py", line 540, in _handle_tasks
    put(task)
  File "/home/gregorio/miniforge3/envs/hetero_dyn/lib/python3.12/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gregorio/miniforge3/envs/hetero_dyn/lib/python3.12/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: cannot pickle 'SwigPyObject' object
```

I know there has been previous discussion about this topic but I wish to ask if there is any development toward being able to run a simple example like this one. The main goal for me is to be able to not have to recreate the optimization program all the time a new sample is given. This should give some performance in finding an optimal solution faster 9without rebuilding the OP all the time). At the same time, the optimization should be parallelised to get more performance (in terms of execution time).

Eventually, this example is part of a bigger project for which the OP will also contain some parameters as `p` 
```
nlpsol("solver","ipopt",{"x":self.x,"f":sin(self.x)**2},"p":p)
```
I would then need to add samples of initial conditions and parameters.

Thank in advance you for any help and suggestions!

**Comment by sandeep026** [comment]
@gregoriomarchesini , 

The workflow to do the task described is different in casadi.
1. try creating the NLP using opti class
2.  use to_function() method to convert the class to a function
3.  call the created function using parameter  and guess. 
4.  use map() method on the function object to solve multple problems in parallel

Refer to [link](https://github.com/casadi/casadi/discussions/3434) for examples on to_function.

casadi objects cannot be pickled. However, the function objects can saved and loaded. (save() and load methods of function class) 

Hope that helps.
Regards
Sandeep

**Reply by sandeep026** [reply]
The map mentioned here is that of casadi's  and does not require the multiprocessing package

**Reply by sandeep026** [reply]
Here is a MWE.

```
import casadi as cs
nlp=cs.Opti()
x=nlp.variable(1)
p=nlp.parameter(1)
obj=(x-p)**2
nlp.minimize(obj)
nlp.solver('ipopt')
f=nlp.to_function('F',[x,p],[x])
P=cs.linspace(1,10,10)
X=10*P
# via horizondtal broadcast/ map('unroll')
R=f(X,P)
print(R)
# parallel eval using threads
F=f.map(10,'thread',2)
R=F(X,P)
print(R)
```

**Reply by gregoriomarchesini** [reply]
Ok thanks a lot for the reply !
I will take a look at the example you provided

Thanks for the support

**Reply by gregoriomarchesini** [reply]
Ok @sandeep026 

I have tried your solution and I can see a speed-up of the code if I increase the threads. The main reason why I wanted to use multi-processing was that I wanted true parallelization and not concurrency.  But probably this is not possible in Casadi

I leave here the code to try changing the thread number and sample size (based on the previous reply of @sandeep026 )

```
import casadi as cs
import time



nlp=cs.Opti()
x=nlp.variable(1)
p=nlp.parameter(1)
obj=(x-p)**2
nlp.minimize(obj)
nlp.solver('ipopt',{"ipopt.print_level":0, "print_time":0})
f=nlp.to_function('F',[x,p],[x])
samples = 100000
thd = 4
P = cs.linspace(1,10,samples)
X = 10*P
# via horizondtal broadcast/ map('unroll')


start = time.time()
for jj in range(samples):
    R=f(X[jj],P[jj])
print("sequential : ", time.time()-start)

start = time.time()
R=f(X,P)
print("single thread : ", time.time()-start)
# parallel eval using threads

start = time.time()
F=f.map(samples,'thread',thd)
R=F(X,P)
print(f"{thd} threads version : ",time.time()-start)

```

Thank you again for your replies and I hope the question can be helpful for someone else also!

**Comment by sandeep026** [comment]
@gregoriomarchesini 

In my experience, if there is a function (with only expressions to be evaluated and not like the NLP ) which is very expensive to evaluate and embarassingly parallel, thread option works and improves evaluation time. If no improvement is seen, it could indicate overheads in parallelization and requires further profiling of the code.

I have modifed my earlier code (just to get started with) to make the NLP higher dimensional and tried to solve 10 instances in serial and parallel. 

```
import casadi as cs
from time import perf_counter
nlp=cs.Opti()
N=50000
x=nlp.variable(N,1)
p=nlp.parameter(N,1)
obj=cs.sumsqr(x-p)
nlp.minimize(obj)
nlp.solver('ipopt',{'ipopt.linear_solver':'spral','ipopt.spral_scaling':'mc64'})
f=nlp.to_function('F',[x,p],[x])
# replace with max logical cpus
M=12
P=cs.repmat(cs.linspace(1,10,N),1,M)
X=10*P
'''
# serial
F=f.map(M,'serial')
a1=perf_counter()
F(X,P)
a2=perf_counter()

# via horizondtal broadcast/ map('unroll')
b1=perf_counter()
f(X,P)
b2=perf_counter()

# python serial
d1=perf_counter()
for i in range(M):
    f(X[:,i],P[:,i])
d2=perf_counter()
'''
# parallel eval using threads
tp=[]
TH=[1,4,8,12]
for i in TH:
    F=f.map(M,'thread',i+1)
    c1=perf_counter()
    R=F(X,P)
    c2=perf_counter()
    tp.append(c2-c1)

#print('serial',a2-a1,'unroll',b2-b1,'thread (max)',c2-c1,'serial python for',d2-d1)

print('thd count','eval. time [s]')
for x,y in dict(zip(TH,tp)).items():
    print(x,y)

```
**N**=10000

If linear solver mumps is used in the given problem, there is no gain performance gain( parallel to serial). If spral is used, there is a huge increase in performance of parallel eval compared to serial. However, mumps works faster in comparison to spral. 

mumps
```
thd count eval. time [s]
1 0.29672569991089404
4 0.30593499983660877
8 0.34049520012922585
12 0.2396607999689877
```

spral

```
thd count eval. time [s]
1 1.9222426998894662
4 1.1282748999074101
8  0.865155499894172
12 0.6118183000944555
```

** N**=50000
For larger problems spral is better suited than mumps.

MUMPS

No improvement in performance

```
thd count eval. time [s]
1 20.9530104000587
4 25.954841300146654
8 31.103365799877793
12 20.72370670014061
```
SPRAL
Almost 4 times faster for 
```
thd count eval. time [s]
1 37.08175110002048
4 21.569982400164008
8 16.384709999896586
12 10.027267999947071
```
